compiler/rustc_hir_typeck/src/method/suggest.rs

```diff
@@ -465,6 +465,48 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             );
         }
 
+        if let Mode::MethodCall = mode && let SelfSource::MethodCall(mut rcvr_expr) = source {
+        while let hir::ExprKind::MethodCall(_path_segment, parent_expr, _args, span) =
+        rcvr_expr.kind
+        {
+            rcvr_expr = parent_expr;
+            let prev_ty = self.resolve_vars_if_possible(
+                self.typeck_results
+                    .borrow()
+                    .expr_ty_adjusted_opt(rcvr_expr)
+                    .unwrap_or(Ty::new_misc_error(self.tcx)),
+            );
+            err.span_note(span, format!("{prev_ty:?}"));
+            println!("prev_ty: {}", prev_ty );
+            println!("item_name: {}", item_name);
+
+            match self.probe_for_name(
+            Mode::MethodCall,
+                item_name,
+                Some(prev_ty),
```

Comment:
This is the problem, this should either be a type that has a valid `.c()` method *or* `None`, but by passing it here we're filtering on all methods that conform to `<T as _>::item_name(&T) -> &T`, which is why we were getting `BadReturnType`. Also, there's `probe_for_name_many` that can give you back multiple methods that might apply and it's more likely what we would want to be using here.

Reply:
Thinking this further, you *want* to pass `None` in 486, but use `probe_for_name_many` instead. This way once we get the result we can look at all the potential places the options could go to suggest an accurate code change, and if none are *positively* valid we emit a `span_note` or `span_label` pointing at the place where the call could go.

Reply:
To be clear, I'm ok with having this PR cover only the "fallback" case and only start providing structured suggestions in a subsequent PR.

Reply:
Been busy at work and will get to this over weekend. Exciting to get to the guts. Thanks for all the pointers!

Reply:
@estebank Spent a few hours on this - I do not understand how `return_type::None` would work to return any function with `foo` `item_name`. Or are we only looking for `foo()` which has a return type of `None`. Are we trying to set `return_type` (for `probe_for_name_namy`) to be the `return_type` of the method with `item_name` in the user's incorrect code : `A.b().c().d().foo()` ?

Reply:
Calling `self.probe_for_name_many(Mode::MethodCall, item_name, None,IsSuggestion(true), prev_ty, rcvr_expr.hir_id, ProbeScope::AllTraits)` instead of `self.probe_for_name_many(Mode::MethodCall, item_name, Some(prev_ty), IsSuggestion(true), prev_ty, rcvr_expr.hir_id, ProbeScope::AllTraits)` will make it so that the machinery looks for methods on all traits of name `item_name` (in this case `foo`) that can be named on `prev_ty`, regardless of return type (that's the `None` we're passing in now).

Before we were explicitly asking for the return type to also be `prev_ty`, which filters on all methods on `prev_ty` that also return `prev_ty`, which there were none and why we were ending up with `BarReturnType`.